chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -118,6 +118,70 @@ impl<I: Interner> ToProgramClauses<I> for AssociatedTyValue<I> {
     }
 }
 
+impl<I: Interner> ToProgramClauses<I> for ImplTraitDatum<I> {
+    /// Given `opaque type T<..>: A + B = HiddenTy;`, we generate:
+    ///
+    /// ```notrust
+    /// AliasEq(T<..> = HiddenTy) :- Reveal.
+    /// AliasEq(T<..> = !T).
+    /// Implemented(!T: A).
+    /// Implemented(!T: B).
+    /// Implemented(!T: Send) :- Implemented(HiddenTy: Send). // For all auto traits
+    /// ```
+    /// where `!T` is the placeholder for the unnormalized type `T<..>`.
+    fn to_program_clauses(&self, builder: &mut ClauseBuilder<'_, I>) {
+        let interner = builder.db.interner();
+        let alias = AliasTy::ImplTrait(ImplTraitTy {
+            impl_trait_id: self.impl_trait_id,
+            substitution: builder.substitution_in_scope(),
+        });
+
+        let alias_ty = Ty::new(interner, alias.clone());
```

Comment:
Hmm, I expected (I think) that we would treat this analogously to other placeholders -- though I've gone a bit back and forth on how that should be. Currently, though, we have a [special variant of `TypeName`](http://rust-lang.github.io/chalk/chalk_ir/enum.TypeName.html#variant.AssociatedType) for the placeholder case. I think maybe we want a variant of `TypeName` also for impl trait placeholders.

Reply:
We want to do this in part because we want unification to treat the placeholder differently (it shouldn't try to prove `AliasEq` if we're using the placeholder, it should only be equal to other placeholders)